cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/ThrowStatement.kt

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2024, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.graph.statements
+
+import de.fraunhofer.aisec.cpg.graph.ArgumentHolder
+import de.fraunhofer.aisec.cpg.graph.edges.ast.astOptionalEdgeOf
+import de.fraunhofer.aisec.cpg.graph.edges.unwrapping
+import de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression
+import java.util.Objects
+import org.neo4j.ogm.annotation.Relationship
+
+/** Represents a `throw` or `raise` statement. */
+class ThrowStatement : Statement(), ArgumentHolder {
+
+    /** The exception object to be raised. */
+    @Relationship(value = "EXCEPTION") var exceptionEdge = astOptionalEdgeOf<Expression>()
+    var exception by unwrapping(ThrowStatement::exceptionEdge)
+
+    /**
+     * Some languages (Python) can add a cause to indicate that an exception was raised while
+     * handling another exception. This is stored in the graph, but has no further implications like
+     * EOG or DFG connections, as it is only of informational purpose, but it doesn't change the
+     * program behavior.
+     */
+    @Relationship(value = "CAUSE") var causeEdge = astOptionalEdgeOf<Expression>()
```

Comment:
Can this only be a `Reference` in all cases? Currently, it's any `Expression`. Adapt the type or it would make sense to include them in the EOG and DFG.

Anyway, I would recommend adding the `cause` to the DFG because e.g. if there's sensitive data inside the Exception in the "cause" we may also have this data in the exception raised here; this could violate some policies and the data flow would be lost without the DFG-edge.

For the EOG, it may be useful as well if we can do some more sophisticated things (and it shouldn't harm to have an edge here I guess).

Reply:
The cause can be any expression, either a call / constructor to a new exception or a reference to an already created one. So, yes it would make sense to have a DFG from cause to throw.

Reply:
Ah, I confused exception with cause... But same applies, in python this can be any expression.

Reply:
Can we maybe rename this field to `parentException`, which is a bit more generic?

Reply:
Renaming done.